internal/singleton/singleton.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package singleton
+
+import (
+	"context"
+	"sync"
+)
+
+// Singleton is a thread-safe type that holds one `T` per process.
+//
+// Example usage:
+//
+//	var globalFoo = singleton.New(singleton.Config[*Foo]{
+//		Constructor: func (ctx context.Context) (*Foo, error) {
+//			return NewFoo(ctx)
+//		},
+//		Destructor: func (ctx context.Context, foo *Foo) error {
+//			return foo.Close(ctx)
+//		})
+//
+// In your setup code:
+//
+//	foo, ticket, err := globalFoo.Acquire(ctx)
+//
+// In your teardown code:
+//
+//	err := globalFoo.Close(ctx, ticket)
+type Singleton[T any] struct {
+	mu         sync.Mutex
+	tickets    map[Ticket]struct{}
+	nextTicket Ticket
+	cfg        Config[T]
+	value      T
+}
+
+// Ticket is an opaque type signifying that a singleton's resource is acquired.
+type Ticket = int
```

Comment:
Why are we using a type alias? In this code I guess it's the same thing as having `type Ticket int`, but the latter would prevent someone from doing `var test int = s.nextTicket` for example. Still doesn't prevent `Close(context.Background(), 42)`, but, alas :)

Reply:
Ah yes I was meaning to define a new type, thanks for catching 😄